lib/util/InternalRepresentations.ts

```diff
@@ -0,0 +1,181 @@
+  };
+}
+
+export function defaultedDateTimeRepresentation(rep: Partial<IDateTimeRepresentation>): IDateTimeRepresentation {
+  return {
+    ...rep,
+    day: rep.day || 1,
+    hours: rep.hours || 0,
+    month: rep.month || 1,
+    year: rep.year || 0,
+    seconds: rep.seconds || 0,
+    minutes: rep.minutes || 0,
+  };
+}
+
+export function toDateTimeRepresentation({ date, timeZone }:
+{ date: Date; timeZone: ITimeZoneRepresentation }): IDateTimeRepresentation {
+  return {
+    year: date.getFullYear(),
+    month: date.getMonth() + 1,
+    day: date.getDate(),
+    hours: date.getHours(),
+    minutes: date.getMinutes(),
+    seconds: date.getSeconds(),
+    zoneHours: timeZone.zoneHours,
+    zoneMinutes: timeZone.zoneMinutes,
+  };
+}
+
+export function convertDurationToDateTime(dur: IDurationRepresentation): IDateTimeRepresentation {
+  return {
+    zoneHours: 0,
+    zoneMinutes: 0,
+    seconds: dur.seconds,
+    minutes: dur.minutes,
+    hours: dur.hours,
+    day: dur.day + 1,
+    month: dur.month + 1,
+    year: dur.year,
+  };
+}
+
+export function negateDuration(dur: Partial<IDurationRepresentation>): Partial<IDurationRepresentation> {
+  return {
+    year: dur.year !== undefined ? -1 * dur.year : undefined,
+    month: dur.month !== undefined ? -1 * dur.month : undefined,
+    day: dur.day !== undefined ? -1 * dur.day : undefined,
+    hours: dur.hours !== undefined ? -1 * dur.hours : undefined,
+    minutes: dur.minutes !== undefined ? -1 * dur.minutes : undefined,
+    seconds: dur.seconds !== undefined ? -1 * dur.seconds : undefined,
+  };
+}
+
+export function toJSDate(date: IDateTimeRepresentation): Date {
+  // The given hours will be assumed to be local time.
+  const res = new Date(
+    date.year,
+    date.month - 1,
+    date.day,
+    date.hours,
+    date.minutes,
+    Math.trunc(date.seconds),
+    (date.seconds % 1) * 1_000,
+  );
+  if (0 <= date.year && date.year < 100) {
+    // Special rule of date has gone int action:
+    // eslint-disable-next-line max-len
+    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date#individual_date_and_time_component_values
+
+    const jumpDeltaOfDate = 1_900;
+    res.setFullYear(res.getFullYear() - jumpDeltaOfDate);
+  }
+  return res;
+}
+
+export function toUTCDate(date: Partial<IDateTimeRepresentation>,
+  defaultTimezone: ITimeZoneRepresentation): Date {
+  const localTime = toJSDate(defaultedDateTimeRepresentation(date));
+  // This date has been constructed in machine local time, now we alter it to become UTC and convert to correct timezone
+  return new Date(
+    localTime.getTime() + (
+      -1 * (date.zoneHours === undefined ? defaultTimezone.zoneHours : date.zoneHours) * 60 -
+      (date.zoneMinutes === undefined ? defaultTimezone.zoneMinutes : date.zoneMinutes) -
+      localTime.getTimezoneOffset()
+    ) * 60 * 1_000,
+  );
+}
+
+export function ensureYearMonthDuration(dur: Partial<IDurationRepresentation>):
```

Comment:
Bad naming. This function ensures your object ONLY contains this. The name might suggest it checks whether it contains this.